### PR TITLE
Capture both Surefire and Failsafe test runs in coverage reports

### PR DESCRIPTION
**Pull request by alzimmermsft**
# Description

Updates the Jacoco configurations in `azure-client-sdk-parent` to allow for both Surefire and Failsafe test runs to be included in the Jacoco code coverage report. Now, in the target folder three Jacoco `exec` files will be created:

- `jacoco-unit.exec` for code coverage captured when running Surefire tests
- `jacoco-it.exec` for code coverage captured when running Failsafe tests
- `jacoco.exec` which is the merged coverage reports for Surefire and Failsafe tests

Code coverage validation is done against `jacoco.exec` as it's the full representation of test coverage.

Additionally, this PR upgrades the versions of `maven-compiler-plugin`, `maven-failsafe-plugin`, and `maven-surefire-plugin` and updates test compilation to ignore the module path, allowing for less difficulties when testing with code not required by the projects source code `module-info`.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/parents/azure-client-sdk-parent/pom.xml [hunk 1]
```diff
@@ -126,8 +126,11 @@
       -XX:+CrashOnOutOfMemoryError
     </heapDumpOnOom>
 
+    <jacocoUnitTestAgent></jacocoUnitTestAgent>
```
Comment (srnagar) [inline sdk/parents/azure-client-sdk-parent/pom.xml:129]: Why do we need this empty tag?

File: sdk/parents/azure-client-sdk-parent/pom.xml [hunk 2]
```diff
@@ -140,8 +143,11 @@
     <additionalSurefireArgLine>
     </additionalSurefireArgLine>
 
+    <jacocoIntegrationTestAgent></jacocoIntegrationTestAgent>
```
Comment (srnagar) [inline sdk/parents/azure-client-sdk-parent/pom.xml:146]: Same here. Why is this empty tag needed?

File: sdk/parents/azure-client-sdk-parent/pom.xml (hunk 1 above)
Comment (alzimmermsft) [reply sdk/parents/azure-client-sdk-parent/pom.xml:129]: Needed by Maven to inject the system property set by Jacoco when preparing the agent, same for the comment below

File: sdk/webpubsub/azure-messaging-webpubsub-client/src/test/java/module-info.java [hunk 3]
```diff
@@ -1,31 +0,0 @@
-// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (mssfang) [inline sdk/webpubsub/azure-messaging-webpubsub-client/src/test/java/module-info.java:1]: webpubsub's module-info java files are deleted? Is this related to the maven surefilre plugin version upgrade?

Comment (alzimmermsft) [reply sdk/webpubsub/azure-messaging-webpubsub-client/src/test/java/module-info.java:1]: These are test code module-infos which we don't really use and cause a lot of problems with build systems and IDEs. And yeah, this was removed due to the compiler plugin changes

Comment (weidongxu-microsoft) [reply sdk/webpubsub/azure-messaging-webpubsub-client/src/test/java/module-info.java:1]: LGTM. Not sure whether they are the reason that before the coverage part of this 2 lib is empty
https://github.com/Azure/azure-sdk-for-java/pull/38335#issuecomment-1899763603

**Comment by alzimmermsft** [comment]
Do not merge until after releases in case the new dependencies cause any issues

File: sdk/appconfiguration/azure-data-appconfiguration/pom.xml [hunk 4]
```diff
@@ -135,38 +135,5 @@
         </dependency>
       </dependencies>
     </profile>
-
-    <profile>
-      <id>java9plus</id>
-      <activation>
-        <jdk>[9,)</jdk>
-      </activation>
```
Comment (srnagar) [inline sdk/appconfiguration/azure-data-appconfiguration/pom.xml:143]: Just curious, why did we have this before for appconfig?

File: sdk/eventgrid/azure-resourcemanager-eventgrid/src/test/java/module-info.java [hunk 5]
```diff
@@ -1,15 +0,0 @@
-// Copyright (c) Microsoft Corporation. All rights reserved.
-// Licensed under the MIT License.
-
-open module com.azure.resourcemanager.eventgrid {
```
Comment (srnagar) [inline sdk/eventgrid/azure-resourcemanager-eventgrid/src/test/java/module-info.java:4]: Do we have live tests running successfully without this module-info.java?

**Review by srnagar — APPROVED** [review]

File: sdk/appconfiguration/azure-data-appconfiguration/pom.xml (hunk 4 above)
Comment (alzimmermsft) [reply sdk/appconfiguration/azure-data-appconfiguration/pom.xml:143]: This resolved issue in test-compile with regards to Java 9+'s module system. With the change to disable the module system in test code this is no longer an issue.

File: sdk/eventgrid/azure-resourcemanager-eventgrid/src/test/java/module-info.java (hunk 5 above)
Comment (alzimmermsft) [reply sdk/eventgrid/azure-resourcemanager-eventgrid/src/test/java/module-info.java:4]: Does Resource Manager have proper live tests? If so, I'll run them

**Pull request merged**